File: app/src/main/java/ru/vegax/xavier/miniMonsterX/IOData/dataLoader/DataLoader.java
package ru.vegax.xavier.miniMonsterX.IOData.dataLoader;

import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Handler;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;

import ru.vegax.xavier.miniMonsterX.IOData.IOFragment;
import ru.vegax.xavier.miniMonsterX.IOData.IOItem;

import static android.os.AsyncTask.Status.FINISHED;

interface CallBack {                   //declare an interface with the callback methods, so you can use on more than one class and just refer to the interface
    void notifyDataChanged();

    void notifyError(String e);

    void setDeviceId(String deviceId);
}

public abstract class DataLoader implements CallBack {
    public static final int PORT_NUMBER = 6;
    private static ArrayList<IOItem> _ioData;
    private static SharedPreferences _preferences;
    private static final String TAG = "DataLoader";
    private static String _urlJSon;
    private static String _urlHtml;
    private static String _urlBase;
    private static boolean _contCyclic; // is continuous cyclic
    private static long _lastRequest;

    private AsyncTask<String, Void, String> _htmDataLoader;
    private AsyncTask<String[], Void, Void> _jsonDataLoader;

    private Handler _handler;
    private Runnable _cyclicRefresh;
    private static int _datauploadRate = 100;

    protected DataLoader(ArrayList<IOItem> ioData, SharedPreferences preferences, String urlJSon, String urlHtml, String urlBase) {
        _urlJSon = urlJSon;
        _urlHtml = urlHtml;
        _urlBase = urlBase;
        _ioData = ioData;
        _preferences = preferences;
        _jsonDataLoader = new DataLoader.GetJSonDataAsync(this).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, new String[]{_urlJSon, _urlHtml});
        // Create the Handler object (on the main thread by default)
        _handler = new Handler();
// Define the code block to be executed
        _cyclicRefresh = () -> {
            // Do something here on the main thread

            if (_contCyclic) {
                refreshData();

            }
            _handler.postDelayed(_cyclicRefresh, _datauploadRate); // Run the above code block on the main thread after 100 milli seconds
        };
        _handler.post(_cyclicRefresh);

    }

    public void stopUptading() {
        _contCyclic = false;
    }

    public void refreshData() {
        _contCyclic = true;
        if (_jsonDataLoader.getStatus() == FINISHED) {
            _jsonDataLoader = new DataLoader.GetJSonDataAsync(this).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, new String[]{_urlJSon, _urlHtml});
        }
    }

    public void setOutputs(String suffix) {
        // _lastRequest = System.currentTimeMillis();
        _htmDataLoader = new SetHtmlDataAsync(this).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, _urlBase + suffix);
    }

    public void cancelTasks() {
        _handler.removeCallbacks(_cyclicRefresh);
        if (_jsonDataLoader != null) {
            _jsonDataLoader.cancel(true);
        }
        if (_htmDataLoader != null) {
            _htmDataLoader.cancel(true);
        }
    }

    public void stopTasks() {
        _contCyclic = false;
        if (_jsonDataLoader != null) {
            _jsonDataLoader.cancel(true);
        }
        if (_htmDataLoader != null) {
            _htmDataLoader.cancel(true);
        }
    }

    private static String getContent(String path, AsyncTask task) throws IOException {
        URL url = new URL(path);
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.connect();
        int timeout = 1;
        con.setConnectTimeout(timeout);// 0.1 seconds
        con.setRequestMethod("GET");
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(con.getInputStream()))) {
            StringBuilder result = new StringBuilder();
            String line;
//            int i = 0;
            while (((line = reader.readLine()) != null) && !task.isCancelled()) {
//                publishProgress(i);
//                i++;
                result.append(line);
            }
//            _progBarDownload.setProgress(100);
            return result.toString();
        }
    }

    public void setConnData(String urlJSon, String urlHtml, String urlBase) {
        _urlJSon = urlJSon;
        _urlHtml = urlHtml;
        _urlBase = urlBase;
    }


    private static class GetJSonDataAsync extends AsyncTask<String[], Void, Void> {
        private final DataLoader _dataLoader;
        private Exception _e = null;
        private String[] _portNames;
        private static boolean[] _masks;
        private boolean[] _states;


        GetJSonDataAsync(DataLoader dataLoader) {
            _dataLoader = dataLoader;
        }

        @Override
        protected Void doInBackground(String[]... urls) {

            String resultUrlJSon;
            String resultUrlHtml;
            try {
                resultUrlJSon = getContent(urls[0][0], this);

                resultUrlHtml = getContent(urls[0][1], this);
                jSonHandleNewData(resultUrlJSon, resultUrlHtml);
            } catch (Exception e) {

                _e = e;
            }

            return null;
        }

        private void jSonHandleNewData(String strJSon, String strHtml) {

            _masks = new boolean[PORT_NUMBER];
            _states = new boolean[PORT_NUMBER];

            try {
                if (!strJSon.contains("}")) {
                    strJSon += "}";
                }
                JSONObject jsonObj = new JSONObject(strJSon);

                // Getting JSON Array node
                String id = jsonObj.getString("id");
                JSONArray statesJs = jsonObj.getJSONArray("prt");
                JSONArray maskJS = jsonObj.getJSONArray("pst");
                _dataLoader.setDeviceId(id);
//                        JSONArray temp = jsonObj.getJSONArray("t");
//                        JSONArray watchdog = jsonObj.getJSONArray("wdr");
//                        String pwm1 = jsonObj.getString("pwm1");
//                        String pwm2 = jsonObj.getString("pwm2");
//                        String pwmt = jsonObj.getString("pwmt");
                for (int i = 0; i < statesJs.length(); i++) {
                    _masks[i] = maskJS.getString(i).equals("1");
                    _states[i] = statesJs.getString(i).equals("1");
                }
                _portNames = getPortNames(strHtml); // take names between [...]

            } catch (final JSONException e) {
                _dataLoader.notifyError(e.getMessage());
            }
        }

        private static String[] getPortNames(String strHtml) {
            String[] portNames = new String[]{"port1", "port2", "port3", "port4", "port5", "port6"};
            int lastStart = 0;
            int lastEnd = 0;
            if (strHtml != null && strHtml.contains("<pre>Manual switch")) {
                for (int i = 0; i < PORT_NUMBER; i++) {
                    int start = strHtml.indexOf('[', lastStart + 1);
                    lastStart = start;

                    int end = strHtml.indexOf(']', lastEnd + 1);
                    lastEnd = end;
                    if (lastStart > 0 && lastEnd > 0) {
                        String helpString = strHtml.substring(start + 1, end);

                        if (_masks != null) {
                            if (_masks[i]) {
                                int braceIndexStart = helpString.indexOf('>');
                                int braceIndexEnd = helpString.indexOf('<', braceIndexStart);
                                portNames[i] = helpString.substring(braceIndexStart + 1, braceIndexEnd);
                            } else {
                                portNames[i] = helpString;
                            }
                        }
                    }
                }
            }

            return portNames;
        }

        @Override
        protected void onPostExecute(Void voids) {
            if (_e == null) {
                populateRecyclerView(_portNames, _masks, _states); // populate reciclerView
                _contCyclic = true;
            } else {
                _contCyclic = false;
                _dataLoader.notifyError(_e.getMessage());
            }
        }

        private void populateRecyclerView(String[] portNames, boolean[] maskJS, boolean[] states) {
            //   _ioData.clear();
            //    if(System.currentTimeMillis()- _lastRequest> _datauploadRate*7){
            for (int i = 0; i < portNames.length; i++) {
                IOItem curItem = null;
                if (_ioData.size() > i) {
                    curItem = _ioData.get(i);
                }
                if (curItem != null) {
                    if (!curItem.isChanging()) {
                        IOItem newItem = new IOItem(portNames[i], maskJS[i], states[i], curItem.isImpulse());
                        _ioData.set(i, newItem);
                    }
                } else {
                    boolean isImpulse = _preferences.getBoolean(IOFragment.PREFF_IMPULSE + _urlBase + i, false);
                    _ioData.add(new IOItem(portNames[i], maskJS[i], states[i], isImpulse));
                }
            }
            _dataLoader.notifyDataChanged();
            //   }
        }
    }

    // sets Html Data but doesn't require response
    private static class SetHtmlDataAsync extends AsyncTask<String, Void, String> {
        private final DataLoader _dataLoader;
        private Exception _e = null;

        SetHtmlDataAsync(DataLoader dataLoader) {
            _dataLoader = dataLoader;
        }

        @Override
        protected String doInBackground(String... urls) {

            String resultUrl = null;
            try {
                resultUrl = getContent(urls[0], this);


            } catch (Exception e) {

                _e = e;
            }
            return resultUrl;
        }

        @Override
        protected void onPostExecute(String strHtml) {
            if (_e != null) {
                _dataLoader.notifyError(_e.getMessage());
            } else {
                _dataLoader.refreshData();

            }
        }

    }
}

File: app/src/main/java/ru/vegax/xavier/minimonsterx/iodata/dataLoader/DataLoader.kt
package ru.vegax.xavier.minimonsterx.iodata.dataLoader

import android.content.SharedPreferences
import android.os.AsyncTask
import android.os.Handler

import org.json.JSONException
import org.json.JSONObject

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.ArrayList

import ru.vegax.xavier.minimonsterx.iodata.IOFragment
import ru.vegax.xavier.minimonsterx.iodata.IOItem

import android.os.AsyncTask.Status.FINISHED

internal interface CallBack {                   //declare an interface with the callback methods, so you can use on more than one class and just refer to the interface
    fun notifyDataChanged()

    fun notifyError(e: String)

    fun setDeviceId(deviceId: String)
}

abstract class DataLoader protected constructor(ioData: ArrayList<IOItem>, preferences: SharedPreferences, urlJSon: String, urlHtml: String, urlBase: String) : CallBack {

    private var mHtmDataLoader: AsyncTask<String, Void, String>? = null
    private var mJsonDataLoader: AsyncTask<Array<String>, Void, Void>? = null


    private val mHandler = Handler()
    private val mCyclicRefresh : Runnable
    init {
        mUrlJSon = urlJSon
        mUrlHtml = urlHtml
        mUrlBase = urlBase
        mIoData = ioData
        mPreferences = preferences
        mJsonDataLoader = DataLoader.GetJSonDataAsync(this).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, arrayOf(mUrlJSon, mUrlHtml))
        // Create the Handler object (on the main thread by default)
        // Define the code block to be executed


        mCyclicRefresh = Runnable{
            if (mContCyclic) {
                refreshData()
            }
            recursiveCall()
        }

        mHandler.post(mCyclicRefresh)

    }
    private fun recursiveCall() {
        mHandler.postDelayed(mCyclicRefresh, mDatauploadRate.toLong()) // Run the above code block on the main thread after 100 milli seconds.
    }

    fun stopUptading() {
        mContCyclic = false
    }

    fun refreshData() {
        mContCyclic = true
        if (mJsonDataLoader!!.status == FINISHED) {
            mJsonDataLoader = DataLoader.GetJSonDataAsync(this).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, arrayOf(mUrlJSon, mUrlHtml))
        }
    }

    fun setOutputs(suffix: String) {
        mHtmDataLoader = SetHtmlDataAsync(this).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, mUrlBase + suffix)
    }

    fun cancelTasks() {
        mHandler.removeCallbacks(mCyclicRefresh)
        if (mJsonDataLoader != null) {
            mJsonDataLoader!!.cancel(true)
        }
        if (mHtmDataLoader != null) {
            mHtmDataLoader!!.cancel(true)
        }
    }

    fun stopTasks() {
        mContCyclic = false
        if (mJsonDataLoader != null) {
            mJsonDataLoader!!.cancel(true)
        }
        if (mHtmDataLoader != null) {
            mHtmDataLoader!!.cancel(true)
        }
    }

    fun setConnData(urlJSon: String, urlHtml: String, urlBase: String) {
        mUrlJSon = urlJSon
        mUrlHtml = urlHtml
        mUrlBase = urlBase
    }


    private class GetJSonDataAsync internal constructor(private val mDataLoader: DataLoader) : AsyncTask<Array<String>, Void, Void>() {
        private var mE: Exception? = null
        private var mPortNames: Array<String>? = null
        private var mStates: BooleanArray? = null

        override fun doInBackground(vararg urls: Array<String>): Void? {

            val resultUrlJSon: String
            val resultUrlHtml: String
            try {
                resultUrlJSon = getContent(urls[0][0], this)

                resultUrlHtml = getContent(urls[0][1], this)
                jSonHandleNewData(resultUrlJSon, resultUrlHtml)
            } catch (e: Exception) {

                mE = e
            }
            return null
        }

        private fun jSonHandleNewData(strJSon: String, strHtml: String) {
            var strMyJSon = strJSon

            mMasks = BooleanArray(PORT_NUMBER)
            mStates = BooleanArray(PORT_NUMBER)

            try {
                if (!strMyJSon.contains("}")) {
                    strMyJSon += "}"
                }
                val jsonObj = JSONObject(strMyJSon)

                // Getting JSON Array node
                val id = jsonObj.getString("id")
                val statesJs = jsonObj.getJSONArray("prt")
                val maskJS = jsonObj.getJSONArray("pst")
                mDataLoader.setDeviceId(id)
                //                        JSONArray temp = jsonObj.getJSONArray("t");
                //                        JSONArray watchdog = jsonObj.getJSONArray("wdr");
                //                        String pwm1 = jsonObj.getString("pwm1");
                //                        String pwm2 = jsonObj.getString("pwm2");
                //                        String pwmt = jsonObj.getString("pwmt");
                for (i in 0 until statesJs.length()) {
                    mMasks[i] = maskJS.getString(i) == "1"
                    mStates!![i] = statesJs.getString(i) == "1"
                }
                mPortNames = getPortNames(strHtml) // take names between [...]

            } catch (e: JSONException) {
                mDataLoader.notifyError(e.message!!)
            }

        }

        override fun onPostExecute(voids: Void?) {
            if (mE == null) {
                populateRecyclerView(mPortNames!!, mMasks, mStates) // populate reciclerView
                mContCyclic = true
            } else {
                mContCyclic = false
                mDataLoader.notifyError(mE!!.message!!)
            }
        }

        private fun populateRecyclerView(portNames: Array<String>, maskJS: BooleanArray?, states: BooleanArray?) {
            //   mIoData.clear();
            //    if(System.currentTimeMillis()- _lastRequest> mDatauploadRate*7){
            for (i in portNames.indices) {
                var curItem: IOItem? = null
                if (mIoData.size > i) {
                    curItem = mIoData[i]
                }
                if (curItem != null) {
                    if (!curItem.isChanging) {
                        val newItem = IOItem(portNames[i], maskJS!![i], states!![i], curItem.isImpulse, false)
                        mIoData[i] = newItem
                    }
                } else {
                    val isImpulse = mPreferences.getBoolean(IOFragment.PREFF_IMPULSE + mUrlBase + i, false)
                    mIoData.add(IOItem(portNames[i], maskJS!![i], states!![i], isImpulse, false))
                }
            }
            mDataLoader.notifyDataChanged()
            //   }
        }

        companion object {
            private lateinit var mMasks: BooleanArray

            private fun getPortNames(strHtml: String?): Array<String> {
                val portNames = arrayOf("port1", "port2", "port3", "port4", "port5", "port6")
                var lastStart = 0
                var lastEnd = 0
                if (strHtml != null && strHtml.contains("<pre>Manual switch")) {
                    for (i in 0 until PORT_NUMBER) {
                        val start = strHtml.indexOf('[', lastStart + 1)
                        lastStart = start

                        val end = strHtml.indexOf(']', lastEnd + 1)
                        lastEnd = end
                        if (lastStart > 0 && lastEnd > 0) {
                            val helpString = strHtml.substring(start + 1, end)

                            if (mMasks[i]) {
                                val braceIndexStart = helpString.indexOf('>')
                                val braceIndexEnd = helpString.indexOf('<', braceIndexStart)
                                portNames[i] = helpString.substring(braceIndexStart + 1, braceIndexEnd)
                            } else {
                                portNames[i] = helpString
                            }
                        }
                    }
                }

                return portNames
            }
        }
    }

    // sets Html Data but doesn't require response
    private class SetHtmlDataAsync internal constructor(private val mDataLoader: DataLoader) : AsyncTask<String, Void, String>() {
        private var mE: Exception? = null

        override fun doInBackground(vararg urls: String): String? {

            var resultUrl: String? = null
            try {
                resultUrl = getContent(urls[0], this)


            } catch (e: Exception) {

                mE = e
            }

            return resultUrl
        }

        override fun onPostExecute(strHtml: String) {
            if (mE != null) {
                mDataLoader.notifyError(mE!!.message!!)
            } else {
                mDataLoader.refreshData()

            }
        }

    }

    companion object {
        const val PORT_NUMBER = 6
        private lateinit var mIoData: ArrayList<IOItem>
        private lateinit var mPreferences: SharedPreferences
        private var mUrlJSon: String = ""
        private var mUrlHtml: String = ""
        private var mUrlBase: String = ""
        private var mContCyclic: Boolean = false // is continuous cyclic
        private val mDatauploadRate = 100

        @Throws(IOException::class)
        private fun getContent(path: String, task: AsyncTask<*, *, *>): String {
            val url = URL(path)
            val con = url.openConnection() as HttpURLConnection
            con.connect()
            val timeout = 1
            con.connectTimeout = timeout// 0.1 seconds
            con.requestMethod = "GET"
            BufferedReader(InputStreamReader(con.inputStream)).use { reader ->
                val result = StringBuilder()
                var line: String?
                line = reader.readLine()
                //            int i = 0;
                while (!task.isCancelled && line != null) {
                    result.append(line)
                    line = reader.readLine()
                }
                return result.toString()
            }
        }
    }
}
